libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,192 @@
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/lsp/ReferenceCollector.h>
+#include <libsolidity/lsp/Utils.h>
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/lsp/LanguageServer.h>
+
+#include <libsolutil/Common.h>
+
+using namespace solidity::frontend;
+using namespace solidity::langutil;
+using namespace std::string_literals;
+using namespace std;
+
+namespace solidity::lsp
+{
+
+namespace
+{
+
+vector<Declaration const*> allAnnotatedDeclarations(Identifier const* _identifier)
+{
+	vector<Declaration const*> output;
+	output.push_back(_identifier->annotation().referencedDeclaration);
+	output += _identifier->annotation().candidateDeclarations;
+	return output;
+}
+
+}
+
+ReferenceCollector::ReferenceCollector(
+	Declaration const& _declaration,
+	std::string const& _sourceIdentifierName
+):
+	m_declaration{_declaration},
+	m_sourceIdentifierName{_sourceIdentifierName.empty() ? _declaration.name() : _sourceIdentifierName}
+{
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	Declaration const* _declaration,
+	ASTNode const& _ast,
+	std::string const& _sourceIdentifierName
+)
+{
+	if (!_declaration)
+		return {};
+
+	ReferenceCollector collector(*_declaration, _sourceIdentifierName);
+	_ast.accept(collector);
+	return std::move(collector.m_resultingReferences);
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	ASTNode const* _sourceNode,
+	SourceUnit const& _sourceUnit
+)
+{
+	if (!_sourceNode)
+		return {};
+
+	auto output = vector<Reference>{};
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_sourceNode))
+	{
+		for (auto const* declaration: allAnnotatedDeclarations(identifier))
+			output += collect(declaration, _sourceUnit, identifier->name());
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(_sourceNode))
+	{
+		solAssert(identifierPath->path().size() >= 1, "");
+		output += collect(identifierPath->annotation().referencedDeclaration, _sourceUnit, identifierPath->path().back());
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(_sourceNode))
+		output += collect(memberAccess->annotation().referencedDeclaration, _sourceUnit, memberAccess->memberName());
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(_sourceNode))
+		output += collect(declaration, _sourceUnit, declaration->name());
+	else
+		lspAssert(false, ErrorCode::InternalError, "Unhandled AST node "s + typeid(*_sourceNode).name());
```

Comment:
`lspAssert` is not an assertion, it's a validation. This should be an actual assert since it's not supposed to ever happen:
```suggestion
		solAssert(false, "Unhandled AST node "s + typeid(*_sourceNode).name());
```

BTW, I think we talked about it and you said you'd rename it because now its name is very confusing and suggest that it's an assertion while it's not (it reports a `RequestError`). Can you do that? You could call it `lspRequire()` or something like that.

Reply:
solAssert is wrong because it would terminate the program. That must be attempted to avoid as much as possible. This is a recoverable error, as only this one RPC call will fail. I can rename the `lspAssert` to `lspRequire` if you want. Not sure though that change should go into **this** PR as it would simply touch way to many other locations. Opinions?

Reply:
Yes, please rename it. And yeah, not in this PR, I'd create a separate one - though it very likely will go into `develop` before this one anyway.

As for recoverability - I think that asserts are fatal enough that they should not be recoverable. And the other way around - if you want to recover from it, then is it really a solid assumption that is guaranteed to be always true in a bug-free program? There are other kinds of errors the LSP will not be able to recover from, like memory violations or out of memory. I consider it this kind of problem. If another, unexpected kind of node appears out of nowhere, the program is just coded wrong, best to shut down ASAP to limit potential damage and just fix the bug.

Reply:
Tracked in #13539